File: ats/src/main/java/fr/automated/trading/systems/pricingmodel/pricingstrategies/neuralnetworks/layer/Layer.java
package fr.automated.trading.systems.pricingmodel.pricingstrategies.neuralnetworks.layer;

import fr.automated.trading.systems.exception.RemoveNeuronOutOfBounds;
import fr.automated.trading.systems.pricingmodel.pricingstrategies.neuralnetworks.neuron.INeuron;
import fr.automated.trading.systems.utils.utils.AtsLogger;

import java.util.ArrayList;

/**
 * Container for neurons. Layer implements ILayer.
 * @author Philippe REMY
 */
public class Layer implements ILayer {

	private final ArrayList<INeuron> neurons = new ArrayList<>();

	public void addNeuron(INeuron neuron) {
		neurons.add(neuron);
	}

	public void removeNeuron(int count) throws RemoveNeuronOutOfBounds {
		
		if(neurons.size() >= count)
		{
			for(int i=0; i<count; i++) {
                neurons.remove(i);
            }
		}
		else
			throw new RemoveNeuronOutOfBounds();
	}

	public INeuron getNeuron(int id) {
		return neurons.get(id);
	}

	public void freeLayer() {
		neurons.removeAll(neurons);
	}

	public void createNewLayer() {
		AtsLogger.log("layer created");
    }

	public int getCount() {
		return neurons.size();
	}

	public Layer() {
		createNewLayer();
	}

}

File: cpp/src/Layer.cpp
#include "../include/Layer.h"
#include <stdlib.h>

using namespace std;

Layer::Layer()
{
    _count = 0;
    _neurons.reserve(20);
    cout << "Layer created" << endl;
}

Layer::~Layer()
{
}

int Layer::addNeuron(Neuron neuron)
{
    _neurons.push_back(neuron);
    cout << "neuron added to layer" << endl;
    return _count++;
}

void Layer::removeNeuron(int count)
{
    if(_count > count)
    {
        while(count > 0)
        {
            _neurons.pop_back();
            _count--;
            count--;
        }
    }
    else
    {
        cout << "Invalid count value : " << count << endl;
        exit(0);
    }
}

Neuron Layer::getNeuron(int id)
{
    return _neurons.at(id);
}

void Layer::freeLayer()
{
    return _neurons.clear();
}

int Layer::getCount()
{
    return _count;
}
